app/assets/javascripts/discourse/app/mixins/textarea-text-manipulation.js

```diff
@@ -322,6 +325,89 @@ export default Mixin.create({
     }
   },
 
+  /**
+   * Removes the provided char from the provided str up
+   * until the limit, or until a character that is _not_
+   * the provided one is encountered.
+   */
+  _deindentLine(str, char, limit) {
+    let eaten = 0;
+    for (let i = 0; i < str.length; i++) {
+      if (eaten < limit && str[i] === char) {
+        eaten += 1;
+      } else {
+        return str.slice(eaten);
+      }
+    }
+    return str;
+  },
+
+  @bind
+  _indentSelection(direction) {
+    if (![INDENT_DIRECTION_LEFT, INDENT_DIRECTION_RIGHT].includes(direction)) {
+      return;
+    }
+
+    const selected = this._getSelected(null, { lineVal: true });
+    const { lineVal } = selected;
+    let value = selected.value;
+
+    // we want to include all the spaces on the selected line as
+    // well, because we want to indent those too, but only if
+    // the user's selection starts at the first character on the
+    // line and not a space. * below indicates the start + end position
+    // of the selection, and the . are spaces:
+    //
+    // YES
+    //
+    //     *
+    // ....text here
+    // ....some more text
+    //                  *
+    //
+    // NO
+    //
+    //   *
+    // ....text here
+    // ....some more text
+    //                  *
```

Comment:
> only if the user's selection starts at the first character on the line and not a space.

I think users can start reporting this behavior as a bug. It's a surprising behavior when you just add one character to selection (or remove one) and indentation stops working.

Usually, text editors just indent all selected lines, no matter at which position the selection starts on the first line. For example, in RubyMine:

https://user-images.githubusercontent.com/1274517/145449772-0acdfb6f-f623-4094-a5dd-bbe0d8236b99.mov

https://user-images.githubusercontent.com/1274517/145449850-3b9eac74-6f26-455d-b756-bb102eaf3742.mov

Sublime text:

https://user-images.githubusercontent.com/1274517/145449880-bcd7b362-9672-4cde-942f-e96e44c5210e.mov

Reply:
Hey you're right, GitHub even does this which is what I was basing things on. I will update the functionality to do this as well.